rust/signed_doc/src/metadata/mod.rs

```diff
@@ -231,33 +232,77 @@ impl Display for Metadata {
     }
 }
 
-impl TryFrom<&Metadata> for coset::Header {
-    type Error = anyhow::Error;
+impl<C> minicbor::Encode<C> for Metadata {
+    fn encode<W: minicbor::encode::Write>(
+        &self, e: &mut minicbor::Encoder<W>, _ctx: &mut C,
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        let number_of_fields = [
+            self.0.content_type.is_some(),
+            self.0.content_encoding.is_some(),
+            self.0.doc_type.is_some(),
+            self.0.id.is_some(),
+            self.0.ver.is_some(),
+            self.extra().doc_ref().is_some(),
+            self.extra().template().is_some(),
+            self.extra().reply().is_some(),
+            self.extra().section().is_some(),
+            !self.extra().collabs().is_empty(),
+            self.extra().parameters().is_some(),
+        ]
+        .iter()
+        .filter(|v| **v)
+        .count();
 
-    fn try_from(meta: &Metadata) -> Result<Self, Self::Error> {
-        let mut builder = coset::HeaderBuilder::new()
-            .content_format(CoapContentFormat::from(meta.content_type()?));
-
-        if let Some(content_encoding) = meta.content_encoding() {
-            builder = builder.text_value(
-                CONTENT_ENCODING_KEY.to_string(),
-                format!("{content_encoding}").into(),
-            );
+        e.map(
+            number_of_fields
+                .try_into()
+                .map_err(minicbor::encode::Error::message)?,
+        )?;
+        if let Some(content_type) = &self.0.content_type {
+            e.encode(3)?.encode(content_type)?;
+        }
+        if let Some(content_encoding) = &self.0.content_encoding {
```

Comment:
The order of these fields is not per the cbor map length first key deterministic encoding rules.